pkg/controllers/workgenerator/controller_integration_test.go

```diff
@@ -0,0 +1,494 @@
+		})
+
+		Context("Test Bound ClusterResourceBinding with a single resource snapshot", func() {
+			var masterSnapshot *fleetv1beta1.ClusterResourceSnapshot
+			var binding *fleetv1beta1.ClusterResourceBinding
+
+			BeforeEach(func() {
+				masterSnapshot = generateResourceSnapshot(1, 1, 0, [][]byte{
+					testClonesetCRD, testNameSpace, testCloneset,
+				})
+				Expect(k8sClient.Create(ctx, masterSnapshot)).Should(Succeed())
+				By(fmt.Sprintf("master resource snapshot  %s created", masterSnapshot.Name))
+				binding = generateClusterResourceBinding(fleetv1beta1.BindingStateBound, masterSnapshot.Name, memberClusterName)
+				Expect(k8sClient.Create(ctx, binding)).Should(Succeed())
+				By(fmt.Sprintf("resource binding  %s created", binding.Name))
+			})
+
+			AfterEach(func() {
+				By("Deleting ClusterResourceBinding")
+				Expect(k8sClient.Delete(ctx, binding)).Should(SatisfyAny(Succeed(), utils.NotFoundMatcher{}))
+				By("Deleting ClusterResourceSnapshot")
+				Expect(k8sClient.Delete(ctx, masterSnapshot)).Should(SatisfyAny(Succeed(), utils.NotFoundMatcher{}))
+			})
+
+			It("Should create the work in the target namespace with master resource snapshot only", func() {
+				// check the work is created
+				work := v1alpha1.Work{}
+				Eventually(func() error {
+					return k8sClient.Get(ctx, types.NamespacedName{Name: testCRPName, Namespace: namespaceName}, &work)
+				}, timeout, interval).Should(Succeed(), "Failed to get the expected work in hub cluster")
+				By(fmt.Sprintf("work %s is created in %s", work.Name, work.Namespace))
+				//inspect the work manifest
+				expectedManifest := []v1alpha1.Manifest{
+					{RawExtension: runtime.RawExtension{Raw: testClonesetCRD}},
+					{RawExtension: runtime.RawExtension{Raw: testNameSpace}},
+					{RawExtension: runtime.RawExtension{Raw: testCloneset}},
+				}
+				diff := cmp.Diff(expectedManifest, work.Spec.Workload.Manifests)
+				Expect(diff).Should(BeEmpty(), fmt.Sprintf("work manifest(%s) mismatch (-want +got):\n%s", work.Name, diff))
+				//inspect the work owner reference
+				expectedOwnerReference := []metav1.OwnerReference{
+					{
+						APIVersion:         fleetv1beta1.GroupVersion.String(),
+						Kind:               "ClusterResourceBinding",
+						Name:               binding.Name,
+						UID:                binding.UID,
+						BlockOwnerDeletion: pointer.Bool(true),
+					},
+				}
+				diff = cmp.Diff(expectedOwnerReference, work.OwnerReferences)
+				Expect(diff).Should(BeEmpty(), fmt.Sprintf("work owner reference(%s) mismatch (-want +got):\n%s", work.Name, diff))
+				//inspect the work labels
+				expectedLabels := map[string]string{
+					fleetv1beta1.CRPTrackingLabel:   testCRPName,
+					fleetv1beta1.ParentBindingLabel: binding.Name,
+				}
+				diff = cmp.Diff(expectedLabels, work.Labels)
+				Expect(diff).Should(BeEmpty(), fmt.Sprintf("work label(%s) mismatch (-want +got):\n%s", work.Name, diff))
```

Comment:
it's better to compare the full structure of the work by ignoring other predictable fields.

https://github.com/golang/go/wiki/TestComments#compare-full-structures

Reply:
@ryanzhang-oss  do you want to address it?

Reply:
I gave it a try but found out that I need to ignore a lot of fields if I compare the entire work such as
metadata.ResourceVersion, UID, Generation, CreationTimestamp, Annotations, ManagedFields.

Here is what the section says.

>  If your struct needs to be compared for approximate equality or some other kind of semantic equality, or it contains fields that cannot be compared for equality (e.g. if one of the fields is an io.Reader), tweaking a [cmp.Diff](https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff) or [cmp.Equal](https://pkg.go.dev/github.com/google/go-cmp/cmp#Equal) comparison with [cmpopts](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts) options such as [cmpopts.IgnoreInterfaces](https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#IgnoreInterfaces) may meet your needs ([example](https://go.dev/play/p/vrCUNVfxsvF)); otherwise, this technique just won't work, so do whatever works.

Reply:
That's true, you have to ignore these runtime fields. The code will become,

```
                                wantWork := v1alpha1.Work{
					ObjectMeta: metav1.ObjectMeta{
						Name:            fmt.Sprintf(fleetv1beta1.FirstWorkNameFmt, testCRPName),
						Namespace:       namespaceName,
						OwnerReferences: expectedOwnerReference,
						Labels:          expectedLabels,
					},
					Spec: v1alpha1.WorkSpec{
						Workload: v1alpha1.WorkloadTemplate{Manifests: expectedManifest},
					},
				}
				ignoreOption := cmpopts.IgnoreFields(metav1.ObjectMeta{},
					"UID", "ResourceVersion", "ManagedFields", "CreationTimestamp", "Generation")
				diff = cmp.Diff(wantWork, work, ignoreOption)
				Expect(diff).Should(BeEmpty(), fmt.Sprintf("work(%s) mismatch (-want +got):\n%s", work.Name, diff))
```
which i think it has better readability. but will leave the decision to you.

Reply:
I think I didn't get it right the first time.